### ISSUE TITLE
source code release

### ISSUE THREAD

**Issue by rofl0r** [brave/brave-browser#2588]
hi, i am having trouble to find the source code release tarball to compile brave myself.
i only found precompiled releases for various linux flavors.
could you please point me into the right direction? thanks!

**Comment by bershan2** [comment]
@rofl0r To my best knowledge, there are no tarballs being published. I'm not sure why you are looking for a tarball, since Brave is a slightly modified version of Chromium, so tarball would contain pretty much all of the Chromium code with a bit of Brave's own code - mostly just UI, HTTPS Everywhere, adblocker, and BAT wallet code.
The instructions for building are [here](https://github.com/brave/brave-browser/wiki#build-instructions). If you need a tarball of the source, just do
```
git clone 

git@github.com:brave/brave-browser.git
cd brave-browser
npm install

# this takes 30-45 minutes to run
# the Chromium source is downloaded which has a large history
npm run init
```
To build it your system would still need the ["Install prerequisites"](https://github.com/brave/brave-browser/wiki#install-prerequisites).

**Comment by rofl0r** [comment]
> I'm not sure why you are looking for a tarball

'tis because i have my own source-based distro, based on musl libc, called [sabotage linux](https://github.com/sabotage-linux/sabotage).

also i want to make it as hard as possible for 3 letter agencies and other thugs to hack into my computer which pretty much excludes pre-built binaries. if they want to hack me, they gotta shelf out a good deal of $$$ to create an exploit that's specifically targetting my own custom build (by which they would need to correctly guess the CFLAGS and all the versions of the library dependencies i use).

i don't think it's unreasonable to ask for a source tarball. other browsers like [firefox](https://github.com/sabotage-linux/sabotage/blob/master/pkg/firefox) do offer the same. unfortunately firefox decided to make `rust` a hard dependency, which is extremely hard to bootstrap, and not portable across all architectures we support, so i've decided to ditch it and look for another browser.

> git clone 

since i want reproducible results and "download-first-build-later", i'd prefer a tarball over having to use git.

> npm install

is there a way without npm ? i really don't want to call another package manager from my distro's package manager.

**Comment by bershan2** [comment]
Disclaimer: I'm not affiliated with Brave, so nothing I say is official. Furthermore, I only recently got interested in Brave (yesterday) because they just [switched to actual Chromium](https://brave.com/brave-upgrades/) (and not Electron fork Muon), as such I don't have much experience with  I'm just writing here in hopes that what I write would be helpful to you.

>is there a way without npm ? i really don't want to call another package manager from my distro's package manager.

Let's start with simpler things: `node` and `npm` are needed just to download and update dependencies, basically to mount other repos somewhere inside of the source tree. So you could run `node` scripts on development machine to download everything and essentially make tarballs yourself (and avoid using `node` and `npm` on users' machines). As far as I understand, `node` and `npm` dependency was introduced during Muon days (because they are part of Electron), but with enough effort they could rewrite all build scripts in this repo in Python (Python is a dependency of Chromium anyway). The build scripts rely on `npm` just for [five modules](https://github.com/brave/brave-browser/blob/master/package-lock.json), which are just libraries for interacting with file system and cli. On the other hand, I don't think anyone cares about rewriting these scripts in Python just to remove a dependency on `node`.

This is repo contains build scripts (so it does not have tarball releases), however, [brave-core does have tarballs](https://github.com/brave/brave-core/releases). Also, [Chromium offers tarballs](https://github.com/chromium/chromium/releases). Have you tried to use those? You would be getting approximately the same experience from either of Chromium and Brave anyway, and Chromium does not use `node` or `npm`. (I assume you are not into BAT and built-in ad blocker, since you mentioned Firefox was your first choice).

>  i want to make it as hard as possible for 3 letter agencies and other thugs to hack into my computer 

On a side note, have you considered signing your commits (to verify integrity of git, as an extra precaution in addition to verifying integrity of packages in package manager)?

**Comment by bershan2** [comment]
@rofl0r Just to reiterate: take a look at brave-core tarballs, may be that's what you need.

**Comment by rofl0r** [comment]
what we have there is not really release tarballs, merely "tarball snapshots from git tags". nonetheless, i downloaded it, but the content is only 35MB, unlikely to contain the full source, so i guess somebody from the @brave-dev team has to step up and shed some light how source users are supposed to properly compile brave **releases** from source.

**Comment by bershan2** [comment]
@rofl0r I don't think Brave publishes complete tarballs that you could just link to, I'm afraid there is a bit more work to this if you don't want to use `node` and `npm`. I'm just suggesting how this could be done.

I still believe building Chromium would be easier (because it simply has fewer dependencies: since Chromium is upstream, every dependency of Chromium is a dependency of Brave, but Brave has few dependencies unique o itself). However, Brave would be a bit more private... But as a first step, I would try to package Chromium in the way you need it. As I mentioned, Chromium provides [tarballs](https://github.com/chromium/chromium/releases). Chromium (and by extension, Brave) uses python and ninja build system (which I [see](https://github.com/sabotage-linux/sabotage/blob/master/pkg/ninja) among your packages), but only Brave uses `node` and `npm`.

As far as I understand, you need a tarball with source (no binary blobs) that could be linked to in a package and be built on users' machines, right?
There are a few challenges to that:
* Practicality: Building Chromium takes a lot of time and resources. E.g.,
    - Time
Building from source and linking is takes a lot of time. Normally, [half an hour](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-dev/PA08hu52-v8) on a server-grade machine with 64GB RAM. On one of my laptops with 8GB of RAM it takes multiple hours. (Fortunately, for developing I can rebuild only a small portion of Brave that depends on files I work on). Release build takes longer due to linking. Is this practical for your users?
    - Disk space
My project build folder `/src/out/Debug` is over 53 GB. (although it is debug build, I still think that during build process all the components get saved to the disk before they are linked. and components themselves are not much smaller).

  Do your users have this kind of motivation? I see you ship llvm-clang, may be it is comparable in size to Chromium/Brave, but I'm not sure about that.

* I don't think Brave offers complete tarballs, however the "tarball snapshots from git tags" that you downloaded are the complete set of changes that Brave makes to Chromium core (but BAT, ad blocker and similar stuff is in separate repos). They apparently don't keep a fork of Chromium project, but instead store the copies of changed files (whole files, not patches) in brave-core and let developers just check out the whole Chromium tree and then copy over Brave's modified files. I think you could just download Chromium tarball, brave-code tarball, copy brave-core files into Chromium folder and build the "Brave-patched" Chromium. Note that you would be effectively just doing what [sync.js --init](https://github.com/brave/brave-browser/blob/master/scripts/sync.js) and [build.js](https://github.com/brave/brave-browser/blob/master/lib/build.js) does right now. Essentially, you could just re-write relevant parts of these files and (a couple of their dependencies) in python/bash/whatever shell you use.

To sum up, I'd recommend doing the following:
1. Check if [Chromium tarballs](https://github.com/chromium/chromium/releases) are OK for your system (and may be, add them to your packages)
2. [Build Brave as recommended.](https://github.com/brave/brave-browser/wiki#build-instructions)
3. If (1) and (2) are successful, rewrite relevant parts of brave-browser JS scripts in python/bash/whatever and in your package link to Chromium tarball and brave-core tarball and use your brave-browser replacement scripts.

**Comment by tomlowenthal** [comment]
@bershan2 is correct: if you want to build Brave from source, you should follow the [build instructions](https://github.com/brave/brave-browser/wiki#build-instructions). I don't understand what your concern is with using git to get the source. Nonetheless, you can download gzipped tarballs of our tags for [`brave-browser`](https://github.com/brave/brave-browser/tags) and [`brave-core`](https://github.com/brave/brave-core/tags) from the tags pages, which are organized under "Releases" on GitHub's "Code" tab.
